pkg/controller/constrainttemplatestatus/constrainttemplatestatus_controller.go

```diff
@@ -0,0 +1,214 @@
+	return add(mgr, r)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(
+	mgr manager.Manager,
+	cs *watch.ControllerSwitch) reconcile.Reconciler {
+	return &ReconcileConstraintStatus{
+		// Separate reader and writer because manager's default client bypasses the cache for unstructured resources.
+		writer:       mgr.GetClient(),
+		statusClient: mgr.GetClient(),
+		reader:       mgr.GetCache(),
+
+		cs:     cs,
+		scheme: mgr.GetScheme(),
+		log:    log,
+	}
+}
+
+var _ handler.Mapper = &Mapper{}
+
+type Mapper struct{}
+
+// Map correlates a ConstraintTemplatePodStatus with its corresponding constraint template
+func (m Mapper) Map(obj handler.MapObject) []reconcile.Request {
+	labels := obj.Meta.GetLabels()
+	name, ok := labels[v1beta1.ConstraintTemplateMapLabel]
+	if !ok {
+		log.Error(fmt.Errorf("constraint template status resource with no mapping label: %s", obj.Meta.GetName()), "missing label while attempting to map a constraint template status resource")
+		return nil
+	}
+	return []reconcile.Request{reconcile.Request{NamespacedName: types.NamespacedName{Name: name}}}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("constraint-template-status-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to ConstraintTemplateStatus
+	err = c.Watch(
+		&source.Kind{Type: &v1beta1.ConstraintTemplatePodStatus{}},
+		&handler.EnqueueRequestsFromMapFunc{ToRequests: &Mapper{}})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to the provided constraint
+	// Watch for changes to ConstraintTemplate
+	err = c.Watch(&source.Kind{Type: &constrainttemplatev1beta1.ConstraintTemplate{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileConstraintStatus{}
+
+// ReconcileSync reconciles an arbitrary constraint object described by Kind
+type ReconcileConstraintStatus struct {
+	reader       client.Reader
+	writer       client.Writer
+	statusClient client.StatusClient
+
+	cs     *watch.ControllerSwitch
+	scheme *runtime.Scheme
+	log    logr.Logger
+}
+
+// +kubebuilder:rbac:groups=constraints.gatekeeper.sh,resources=*,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=status.gatekeeper.sh,resources=*,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile reads that state of the cluster for a constraint object and makes changes based on the state read
+// and what is in the constraint.Spec
+func (r *ReconcileConstraintStatus) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Short-circuit if shutting down.
+	if r.cs != nil {
+		running := r.cs.Enter()
+		defer r.cs.Exit()
+		if !running {
+			return reconcile.Result{}, nil
+		}
+	}
+	template := &unstructured.Unstructured{}
```

Comment:
I'm split on whether this should be changed. On one hand, duplicating RAM usage is bad. On the other, this allows us to fully decouple the schema of the status resource from the schema as stored in the constraint framework, which means we can change our status reporting without being blocked on updating the framework.

Reply:
> which means we can change our status reporting without being blocked on updating the framework
Is that true? Just because we're writing unstructured data doesn't mean there isn't an underlying schema we should be adhering to.

Personally, I'd go with the structured approach if for nothing else then for consistency in how we interact with ConstraintTemplates in other parts of the code.

Reply:
Status in the constraint framework was always a "meh" decision. Because the framework is meant to be platform neutral, status that makes sense for one integration may not make sense for a different one. It's not so much that there isn't an underlying schema, as that underlying schema should not be defined by the constraint framework.

I suppose we could replace the strongly-typed constraint template status resource with a `map[string]interface{}` resource and treat it similarly to constraints. We could even create our own schema and serialize to/from it if we wanted to introduce strong typing at the constraint framework level, WDYT?

Reply:
Sounds like something to consider in a separate PR? :)

Reply:
SGTM, filed issue #656